mrmustard/lab/abstract/state.py

```diff
@@ -394,6 +345,120 @@ def primal(self, other: Union[State, Transformation]) -> State:
                 f"Cannot apply {other.__class__.__qualname__} to {self.__class__.__qualname__}"
             ) from e
 
+    def _project_onto_state(self, other: State) -> Union[State, float]:
+        """If states are gaussian use generaldyne measurement, else use
+        the states' Fock representation."""
+
+        # if both states are gaussian
+        if self.is_gaussian and other.is_gaussian:
+            return self._project_onto_gaussian(other)
+
+        # either self or other is not gaussian
+        return self._project_onto_fock(other)
+
+    def _project_onto_fock(self, other: State) -> Union[State, float]:
```

Comment:
I feel this method is in the wrong place. It should be in the physics module, because otherwise you have to import from lab (this messes with dependency inversion).

Reply:
`Homodyne` is here just for typechecking. I have removed it in favour of not imprting anything from lab, however the attribute `sample` will come only from `Homodyne` instances. The latter implies that there is a hidden dependency anyhow.

I left a comment in the code so it is easy to track in the future.